application/app/Http/Controllers/DepartmentController.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Http\Requests\StoreDepartmentRequest;
+use App\Http\Requests\UpdateDepartmentRequest;
+use App\Http\Resources\DepartmentResource;
+use App\Models\Department;
+use App\Models\Institution;
+use App\Policies\DepartmentPolicy;
+use Illuminate\Auth\Access\AuthorizationException;
+use Illuminate\Contracts\Database\Eloquent\Builder;
+use Illuminate\Http\Request;
+use Illuminate\Http\Resources\Json\ResourceCollection;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\DB;
+use Throwable;
+
+class DepartmentController extends Controller
+{
+    public function index(): ResourceCollection
+    {
+        return DepartmentResource::collection($this->getBaseQuery()->get());
+    }
+
+    /**
+     * @throws AuthorizationException
+     * @throws Throwable
+     */
+    public function store(StoreDepartmentRequest $request): DepartmentResource
+    {
+        return DB::transaction(function () use ($request): DepartmentResource {
+            $this->authorize('create', Department::class);
+
+            $currentInstitution = Institution::findOrFail(Auth::user()->institutionId);
+
+            $newDepartment = new Department($request->validated());
+            $newDepartment->institution()->associate($currentInstitution);
+            $newDepartment->saveOrFail();
+
+            return new DepartmentResource($newDepartment->refresh());
+        });
+    }
+
+    /**
+     * @throws AuthorizationException
+     */
+    public function show(Request $request): DepartmentResource
+    {
+        $id = $request->route('department_id');
+        $department = $this->getBaseQuery()->findOrFail($id);
+
+        $this->authorize('view', $department);
```

Comment:
[non-blocking] Just a minor suggestion to move the authorization check before querying a department 🙂

Reply:
@kadmit Hmm, unfortunately I need the department object in order to perform the authorization check: the authorization method checks that the selected department’s institution matches the acting user’s institution.

![image](https://github.com/keeleinstituut/tv-authorization/assets/26142658/1d8fdd66-6ed7-42cf-8a72-07ca0975f024)

If you see another way of how we could move the authorization check before querying – let me know. :)

Reply:
@DanBondarenko Yes, that's correct, my mistake 🙂